### PR TITLE
#A09. As Admin, I would like to trigger an event to award the winning project(s) based on the judges scores.

### PR DESCRIPTION
**Pull request by stefdworschak**
## Description
- Adding functionality to create hackathon awards for hackathon automatically upon creation (still need to implement a view to edit them)
- Changing the score category model to Many-to-Many to allow adding them to multiple hackathons. Also adding them to the hackathon create/update form.
- Updating score calculation model in order for the scores to render always event if not all judges have scored yet. (still TBD if we should exclude judges' scores if they have not scored all teams)

## Pull request type

- [ ] #M01 (Miscellaneous User Story #01)
- [ ] #P02 (Participant User Story #02)
- [ ] #S03 (Staff User Story #03)
- [x] #A04 (Admin User Story #04)
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue
#133

## Configuration instructions
- Run migrations

## Testing
All tests are passing.

## Screenshots
Screenshot from 2021-01-20 11-46-32

### CODE REVIEW

File: hackathon/views.py [hunk 1]
```diff
@@ -109,37 +122,113 @@ def check_projects_scores(request, hack_id):
...
+        team_scores = {}
+        category_scores_per_team = {}
+        score_categories = HackProjectScoreCategory.objects.all()
+        hackathon_projects = [team.project.id for team in hackathon.teams.all()
+                              if team.project]
+        scores = HackProjectScore.objects.filter(
+            project_id__in=hackathon_projects).all()
+        judges = hackathon.judges.all()
+
+        # Creating scores data structure
```
Comment (TravelTimN) [inline hackathon/views.py:151]: Do we need this in here still? All commented-out for a reason (keeping for later?)

File: hackathon/views.py [hunk 2]
```diff
@@ -226,6 +328,9 @@ def update_hackathon(request, hackathon_id):
         if form.is_valid():
             form.instance.updated = now
             form.save()
+            # saved_form = form.save(commit=False)
```
Comment (TravelTimN) [inline hackathon/views.py:331]: Do we need these still if commented-out?

File: hackathon/views.py [hunk 3]
```diff
@@ -327,3 +432,13 @@ def enroll_toggle(request):
...
+
+
+# @login_required
+# def update_hack_awards(request):
+#     if request.method == 'POST':
+#         pass
+#     else:
+#         HackAwardCategoryFormSet = modelformset_factory(
+#             HackAwardCategory, fields=('display_name', 'project'),
+#             formset=HackAwardCategoryFormSet)
```
Comment (TravelTimN) [inline hackathon/views.py:444]: Missing blank line at end of file....
But do we need this last function `def update_hack_awards(request):` if it's all commented-out anyway?

File: left_to_do.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+- Discuss if we want to exclude judges who have not scored all teams from the scoring
```
Comment (TravelTimN) [inline left_to_do.txt:1]: Yep, I think we should (if time permits) only calculate Judges scores for those Judges that have supplied scores to each team in the particular Hackathon, not any that are left-out.

File: static/css/style.css [hunk 5]
```diff
@@ -227,6 +229,15 @@ html .all-auth-container {
     color: var(--p-orange);
 }
 
+.enroll-hackathon-card, .enroll-hackathon-card form {
+    width: fit-content;
+    display: inline-block;
+}
+.enroll-hackathon-card form button#enroll-part {
+    margin: 0px!important;
```
Comment (TravelTimN) [inline static/css/style.css:237]: `margin: 0 !important;`
no `px` needed, and space required

**Review by TravelTimN — APPROVED** [review]
A few notes, but apart from those, LGTM and really great work here! 🤩

File: hackathon/fixtures/hackathons.json [hunk 6]
```diff
@@ -275,5 +275,83 @@
...
+            "created": "2021-01-18T23:27:06.554Z",
+            "updated": "2021-01-18T23:27:06.554Z",
+            "created_by": 1,
+            "display_name": "Most Creative Project",
+            "description": "The project with the most creative idea.",
+            "hackathon": 4,
+            "winning_project": null
+        }
     }
 ]
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/fixtures/hackathons.json:357]: Added blank line.

File: hackathon/forms.py [hunk 7]
```diff
@@ -104,3 +113,24 @@ class Meta:
...
+        required=True
+    )
+
+    class Meta:
+        model = HackAwardCategory
+        fields = ('id', 'display_name', 'winning_project')
+
+    def __init__(self, *args, **kwargs):
+        super(HackAwardCategoryForm, self).__init__(*args, **kwargs)
+        self.fields['display_name'].widget.attrs['readonly'] = True
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/forms.py:136]: Added blank line.

File: hackathon/views.py (hunk 1 above)
Comment (stefdworschak) [reply hackathon/views.py:151]: This is just so other developers have a picture of what the data structure will look like.

File: hackathon/views.py (hunk 2 above)
Comment (stefdworschak) [reply hackathon/views.py:331]: Removed commented out code.

File: hackathon/views.py (hunk 3 above)
Comment (stefdworschak) [reply hackathon/views.py:444]: Not for this PR, will add another PR to add this view.

File: static/css/style.css (hunk 5 above)
Comment (stefdworschak) [reply static/css/style.css:237]: Fixed css as suggested.

File: left_to_do.txt (hunk 4 above)
Comment (stefdworschak) [reply left_to_do.txt:1]: Cool. Will add a follow-up issue and PR for this.

**Pull request merged**